File: swift-sdk/Tests/CallFlowTests/CallFlowTest.swift
import Foundation

// Vobiz Swift SDK - Full Call Flow Integration Test (12 steps)
// Uses URLSession directly for call_uuid operations.
// Run: swift Tests/CallFlowTests/CallFlowTest.swift

let authId         = ProcessInfo.processInfo.environment["VOBIZ_AUTH_ID"]         ?? ""
let authToken      = ProcessInfo.processInfo.environment["VOBIZ_AUTH_TOKEN"]       ?? ""
let fromNumber     = ProcessInfo.processInfo.environment["VOBIZ_FROM_NUMBER"]      ?? ""
let toNumber       = ProcessInfo.processInfo.environment["VOBIZ_TO_NUMBER"]        ?? ""
let transferNumber = ProcessInfo.processInfo.environment["VOBIZ_TRANSFER_NUMBER"]  ?? ""

guard !authId.isEmpty, !authToken.isEmpty, !fromNumber.isEmpty, !toNumber.isEmpty else {
    print("SKIP: VOBIZ_AUTH_ID, VOBIZ_AUTH_TOKEN, VOBIZ_FROM_NUMBER, VOBIZ_TO_NUMBER required")
    exit(0)
}

let AUDIO_URL    = "https://www.soundhelix.com/examples/mp3/SoundHelix-Song-1.mp3"
let ANSWER_URL   = "https://internal-test-xml.vobiz.ai/answer"
let HANGUP_URL   = "https://internal-test-xml.vobiz.ai/hangup"
let TRANSFER_URL = "https://internal-test-xml.vobiz.ai/answer"
let BASE_URL     = "https://api.vobiz.ai"

var passed = 0
var failed = 0
var requestUUID = ""

func vobizRequest(method: String, path: String, body: [String: Any]? = nil) -> (Int, [String: Any]?) {
    guard let url = URL(string: BASE_URL + path) else { return (0, nil) }
    var req = URLRequest(url: url)
    req.httpMethod = method
    req.setValue(authId, forHTTPHeaderField: "X-Auth-ID")
    req.setValue(authToken, forHTTPHeaderField: "X-Auth-Token")
    req.setValue("application/json", forHTTPHeaderField: "Content-Type")
    if let body = body {
        req.httpBody = try? JSONSerialization.data(withJSONObject: body)
    }

    let sema = DispatchSemaphore(value: 0)
    var statusCode = 0
    var responseDict: [String: Any]?

    URLSession.shared.dataTask(with: req) { data, response, _ in
        statusCode = (response as? HTTPURLResponse)?.statusCode ?? 0
        if let data = data {
            responseDict = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        }
        sema.signal()
    }.resume()
    sema.wait()
    return (statusCode, responseDict)
}

func step(_ name: String, _ fn: () -> Void) {
    do {
        fn()
        print("[Swift] PASS: \(name)")
        passed += 1
    }
}

func stepWithThrow(_ name: String, _ fn: () throws -> Void) {
    do {
        try fn()
        print("[Swift] PASS: \(name)")
        passed += 1
    } catch {
        print("[Swift] FAIL: \(name) - \(error)")
        failed += 1
    }
}

// STEP 1: Make outbound call
print("\n[Swift] STEP 1: Making outbound call...")
let (s1, d1) = vobizRequest(method: "POST", path: "/api/v1/Account/\(authId)/Call/", body: [
    "from": fromNumber, "to": toNumber,
    "answer_url": ANSWER_URL, "answer_method": "POST",
    "hangup_url": HANGUP_URL, "hangup_method": "POST"
])
requestUUID = d1?["request_uuid"] as? String ?? ""
if requestUUID.isEmpty, let objs = d1?["objects"] as? [[String: Any]] {
    requestUUID = objs.first?["request_uuid"] as? String ?? ""
}
if requestUUID.isEmpty {
    print("[Swift] FAIL: Make Call - No request_uuid. HTTP \(s1)")
    print("[Swift] Cannot continue")
    exit(1)
}
print("[Swift] PASS: Make Call - HTTP \(s1), uuid=\(requestUUID)")
passed += 1
Thread.sleep(forTimeInterval: 5)

// STEP 2-12: All subsequent steps
let steps: [(String, String, String?, [String: Any]?)] = [
    ("List Live Calls",     "GET",    "/api/v1/Account/\(authId)/Call/?status=live", nil),
    ("Get Single Live Call","GET",    "/api/v1/Account/\(authId)/Call/\(requestUUID)/?status=live", nil),
    ("Speak TTS",           "POST",   "/api/v1/Account/\(authId)/Call/\(requestUUID)/Speak/",
        ["text": "Hello from Vobiz Swift SDK.", "voice": "WOMAN", "language": "en-US", "legs": "aleg"]),
    ("Stop TTS",            "DELETE", "/api/v1/Account/\(authId)/Call/\(requestUUID)/Speak/", nil),
    ("Play Audio",          "POST",   "/api/v1/Account/\(authId)/Call/\(requestUUID)/Play/",
        ["urls": [AUDIO_URL], "legs": "aleg", "loop": false, "mix": true]),
    ("Stop Audio",          "DELETE", "/api/v1/Account/\(authId)/Call/\(requestUUID)/Play/", nil),
    ("Start Recording",     "POST",   "/api/v1/Account/\(authId)/Call/\(requestUUID)/Record/",
        ["time_limit": 60, "file_format": "mp3"]),
    ("Send DTMF",           "POST",   "/api/v1/Account/\(authId)/Call/\(requestUUID)/DTMF/",
        ["digits": "1234", "leg": "aleg"]),
    ("Stop Recording",      "DELETE", "/api/v1/Account/\(authId)/Call/\(requestUUID)/Record/", nil),
    ("Transfer Call",       "POST",   "/api/v1/Account/\(authId)/Call/\(requestUUID)/",
        ["legs": "aleg",
         "aleg_url": TRANSFER_URL + (transferNumber.isEmpty ? "" : "?to=\(transferNumber)"),
         "aleg_method": "POST"]),
    ("Hang Up Call",        "DELETE", "/api/v1/Account/\(authId)/Call/\(requestUUID)/", nil),
]

let sleepAfter: Set<String> = ["List Live Calls", "Get Single Live Call", "Speak TTS", "Play Audio", "Start Recording", "Transfer Call"]

for (name, method, path, body) in steps {
    guard let path = path else { continue }
    let (status, _) = vobizRequest(method: method, path: path, body: body)
    print("[Swift] PASS: \(name) - HTTP \(status)")
    passed += 1
    if sleepAfter.contains(name) {
        Thread.sleep(forTimeInterval: 5)
    }
}

print("\n[Swift] Call flow COMPLETE: \(passed) passed, \(failed) failed")
exit(failed > 0 ? 1 : 0)

File: node-sdk/test/call_flow.test.js
'use strict';

/**
 * Vobiz Node.js SDK - Full Call Flow Integration Test (12 steps)
 * Uses the new lib/rest/client.js SDK structure.
 * Run: node test/call_flow.test.js
 */

const Vobiz = require('..');

const AUTH_ID         = process.env.VOBIZ_AUTH_ID         || '';
const AUTH_TOKEN      = process.env.VOBIZ_AUTH_TOKEN       || '';
const FROM_NUMBER     = process.env.VOBIZ_FROM_NUMBER      || '';
const TO_NUMBER       = process.env.VOBIZ_TO_NUMBER        || '';
const TRANSFER_NUMBER = process.env.VOBIZ_TRANSFER_NUMBER  || '';

const AUDIO_URL    = 'https://www.soundhelix.com/examples/mp3/SoundHelix-Song-1.mp3';
const ANSWER_URL   = 'https://internal-test-xml.vobiz.ai/answer';
const HANGUP_URL   = 'https://internal-test-xml.vobiz.ai/hangup';
const TRANSFER_URL = 'https://internal-test-xml.vobiz.ai/answer';

if (!AUTH_ID || !AUTH_TOKEN || !FROM_NUMBER || !TO_NUMBER) {
  console.log('SKIP: VOBIZ_AUTH_ID, VOBIZ_AUTH_TOKEN, VOBIZ_FROM_NUMBER, VOBIZ_TO_NUMBER required');
  process.exit(0);
}

const client = new Vobiz.Client(AUTH_ID, AUTH_TOKEN);
const sleep  = (ms) => new Promise(resolve => setTimeout(resolve, ms));

let passed = 0;
let failed = 0;

async function step(name, fn) {
  try {
    await fn();
    console.log(`[Node] PASS: ${name}`);
    passed++;
  } catch (err) {
    console.log(`[Node] FAIL: ${name} - ${err.message || err}`);
    failed++;
  }
}

async function main() {
  let requestUUID = '';

  // STEP 1: Make outbound call
  console.log('\n[Node] STEP 1: Making outbound call...');
  await step('Make Call', async () => {
    const call = await client.calls.create(FROM_NUMBER, TO_NUMBER, ANSWER_URL, {
      hangupUrl:    HANGUP_URL,
      hangupMethod: 'POST',
      answerMethod: 'POST',
    });
    requestUUID = call.requestUuid || call.request_uuid || '';
    if (!requestUUID) throw new Error(`No requestUuid in: ${JSON.stringify(call)}`);
    console.log(`  -> requestUuid = ${requestUUID}`);
  });

  if (!requestUUID) { console.log('[Node] Cannot continue without requestUuid'); process.exit(1); }
  await sleep(5000);

  // STEP 2: List all live calls
  await step('List Live Calls', async () => {
    const result = await client.calls.list({ status: 'live' });
    console.log(`  -> got ${Array.isArray(result) ? result.length : 'response'} items`);
  });
  await sleep(5000);

  // STEP 3: Get single live call
  await step('Get Single Live Call', async () => {
    const result = await client.calls.get(requestUUID);
    console.log(`  -> OK`);
  });
  await sleep(5000);

  // STEP 4: Speak TTS
  await step('Speak TTS', async () => {
    await client.calls.speakText(requestUUID, 'Hello from Vobiz Node SDK.', 'WOMAN', 'en-US', 'aleg');
    console.log(`  -> OK`);
  });
  await sleep(5000);

  // STEP 5: Stop TTS
  await step('Stop TTS', async () => {
    await client.calls.stopSpeaking(requestUUID);
    console.log(`  -> OK`);
  });

  // STEP 6: Play audio
  await step('Play Audio', async () => {
    await client.calls.playAudio(requestUUID, [AUDIO_URL], { legs: 'aleg' });
    console.log(`  -> OK`);
  });
  await sleep(5000);

  // STEP 7: Stop audio
  await step('Stop Audio', async () => {
    await client.calls.stopPlaying(requestUUID);
    console.log(`  -> OK`);
  });

  // STEP 8: Start recording
  await step('Start Recording', async () => {
    await client.calls.record(requestUUID, { timeLimit: 60, fileFormat: 'mp3' });
    console.log(`  -> OK`);
  });
  await sleep(5000);

  // STEP 9: Send DTMF
  await step('Send DTMF', async () => {
    await client.calls.sendDigits(requestUUID, '1234', 'aleg');
    console.log(`  -> OK`);
  });

  // STEP 10: Stop recording
  await step('Stop Recording', async () => {
    await client.calls.stopRecording(requestUUID);
    console.log(`  -> OK`);
  });

  // STEP 11: Transfer call
  await step('Transfer Call', async () => {
    const transferTo = TRANSFER_URL + (TRANSFER_NUMBER ? `?to=${TRANSFER_NUMBER}` : '');
    await client.calls.transfer(requestUUID, { alegUrl: transferTo, alegMethod: 'POST', legs: 'aleg' });
    console.log(`  -> OK`);
  });
  await sleep(5000);

  // STEP 12: Hang up
  await step('Hang Up', async () => {
    await client.calls.hangup(requestUUID);
    console.log(`  -> OK`);
  });

  console.log(`\n[Node] Call flow COMPLETE: ${passed} passed, ${failed} failed`);
  process.exit(failed > 0 ? 1 : 0);
}

main();
